kernel/bpf/helpers.c

```diff
@@ -3702,6 +3704,195 @@ __bpf_kfunc int bpf_strstr(const char *s1__ign, const char *s2__ign)
+
+	/*
+	 * If map element is freed, context state is set to BPF_TW_FREED
+	 * Otherwise it is safe to access map key value under the rcu_read_lock
+	 */
+	rcu_read_lock_trace();
+	state = cmpxchg(&ctx->state, BPF_TW_SCHEDULING, BPF_TW_RUNNING);
+	if (state == BPF_TW_SCHEDULED)
+		state = cmpxchg(&ctx->state, BPF_TW_SCHEDULED, BPF_TW_RUNNING);
+	if (state == BPF_TW_FREED) {
+		rcu_read_unlock_trace();
+		bpf_reset_task_work_context(ctx);
+		goto free;
+	}
+
+	key = (void *)map_key_from_value(ctx->map, ctx->map_val, &idx);
+	migrate_disable();
+	ctx->callback_fn(ctx->map, key, ctx->map_val);
+	migrate_enable();
+
+	rcu_read_unlock_trace();
+	/* State is running or freed, either way reset. */
+	bpf_reset_task_work_context(ctx);
+	state = cmpxchg(&ctx->state, BPF_TW_RUNNING, BPF_TW_STANDBY);
+	if (state == BPF_TW_RUNNING)
+		return; /* Work is done, context is reset */
+
+free:
+	kfree(ctx);
+}
+
+static void bpf_task_work_irq(struct irq_work *irq_work)
+{
+	struct bpf_task_work_context *ctx;
+	enum bpf_task_work_state state;
+	int err;
+
+	ctx = container_of(irq_work, struct bpf_task_work_context, irq_work);
+
+	state = cmpxchg(&ctx->state, BPF_TW_PENDING, BPF_TW_SCHEDULING);
+	if (state == BPF_TW_FREED)
+		goto free_context;
+
+	err = task_work_add(ctx->task, &ctx->work, ctx->mode);
+	if (err) {
+		bpf_reset_task_work_context(ctx);
+		state = cmpxchg(&ctx->state, BPF_TW_SCHEDULING, BPF_TW_STANDBY);
+		if (state == BPF_TW_FREED)
+			kfree(ctx);
+		return;
+	}
+	state = cmpxchg(&ctx->state, BPF_TW_SCHEDULING, BPF_TW_SCHEDULED);
+	if (state == BPF_TW_FREED &&
+	    task_work_cancel_match(ctx->task, task_work_match, ctx))
+		goto free_context; /* successful cancellation, release and free ctx */
+	return;
+
+free_context:
+	bpf_reset_task_work_context(ctx);
+	kfree(ctx);
+}
+
+static struct bpf_task_work_context *
+bpf_task_work_aquire_ctx(struct bpf_task_work *tw, struct bpf_map *map)
+{
+	struct bpf_task_work_context *ctx;
+	enum bpf_task_work_state state;
+	atomic_long_t *ctx_ptr = (atomic_long_t *)&tw->ctx;
+
+	ctx = (void *)atomic_long_read(ctx_ptr);
+	if (!ctx) {
+		ctx = bpf_map_kmalloc_node(map,
+					   sizeof(struct bpf_task_work_context),
+					   GFP_ATOMIC, map->numa_node);
+		if (!ctx)
+			return ERR_PTR(-ENOMEM);
+		memset(ctx, 0, sizeof(*ctx));
+		if (atomic_long_cmpxchg(ctx_ptr, 0, (unsigned long)ctx) != 0) {
+			kfree(ctx);
+			return ERR_PTR(-EBUSY);
+		}
+	}
+	state = cmpxchg(&ctx->state, BPF_TW_STANDBY, BPF_TW_PENDING);
+	if (state != BPF_TW_STANDBY)
+		return ERR_PTR(-EBUSY);
+
+	return ctx;
+}
+
+static int bpf_task_work_schedule(struct task_struct *task, struct bpf_task_work *tw,
+				  struct bpf_map *map, bpf_task_work_callback_t callback_fn,
+				  struct bpf_prog_aux *aux, enum task_work_notify_mode mode)
+{
+	struct bpf_prog *prog;
+	struct bpf_task_work_context *ctx = NULL;
+	int err;
+
+	BTF_TYPE_EMIT(struct bpf_task_work);
+
+	prog = bpf_prog_inc_not_zero(aux->prog);
+	if (IS_ERR(prog))
+		return -EPERM;
```

Comment:
EPERM doesn't sound right, this has nothing to do with permissions... Not sure what's the best match, but maybe -EBADf or -ENXIO?..